src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHelper.cs

```diff
@@ -0,0 +1,123 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal static class DiagnosticsHelper
+    {
+        internal static string GetRedactedUriString(Uri uri)
+        {
+            Debug.Assert(uri.IsAbsoluteUri);
+
+            if (GlobalHttpSettings.DiagnosticsHandler.DisableUriRedaction)
+            {
+                return uri.AbsoluteUri;
+            }
+
+            string pathAndQuery = uri.PathAndQuery;
+            int queryIndex = pathAndQuery.IndexOf('?');
+
+            bool redactQuery = queryIndex >= 0 && // Query is present.
+                queryIndex < pathAndQuery.Length - 1; // Query is not empty.
```

Comment:
We're getting the path and query but then trying to strip off the query? Why not just get the path? Is it that we're trying to only append a * when something was redacted / avoid adding a * if nothing was removed?

Reply:
> Is it that we're trying to only append a * when something was redacted / avoid adding a * if nothing was removed?

Yes, `*` signals to the user that redaction happened.

Reply:
Using `PathAndQuery` is also currently cheaper because we'll cache the allocated string that we'll need when sending the request anyway.